datafusion/spark/src/function/math/hex.rs

```diff
@@ -0,0 +1,404 @@
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::any::Any;
+use std::sync::Arc;
+
+use crate::function::error_utils::{
+    invalid_arg_count_exec_err, unsupported_data_type_exec_err,
+};
+use arrow::array::{Array, StringArray};
+use arrow::datatypes::DataType;
+use arrow::{
+    array::{as_dictionary_array, as_largestring_array, as_string_array},
+    datatypes::Int32Type,
+};
+use datafusion_common::{
+    cast::{as_binary_array, as_fixed_size_binary_array, as_int64_array},
+    exec_err, DataFusionError,
+};
+use datafusion_expr::Signature;
+use datafusion_expr::{ColumnarValue, ScalarFunctionArgs, ScalarUDFImpl, Volatility};
+use std::fmt::Write;
+
+/// <https://spark.apache.org/docs/latest/api/sql/index.html#hex>
+#[derive(Debug)]
+pub struct SparkHex {
+    signature: Signature,
+    aliases: Vec<String>,
+}
+
+impl Default for SparkHex {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl SparkHex {
+    pub fn new() -> Self {
+        Self {
+            signature: Signature::user_defined(Volatility::Immutable),
+            aliases: vec![],
+        }
+    }
+}
+
+impl ScalarUDFImpl for SparkHex {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn name(&self) -> &str {
+        "hex"
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(
+        &self,
+        _arg_types: &[DataType],
+    ) -> datafusion_common::Result<DataType> {
+        Ok(DataType::Utf8)
+    }
+
+    fn invoke_with_args(
+        &self,
+        args: ScalarFunctionArgs,
+    ) -> datafusion_common::Result<ColumnarValue> {
+        spark_hex(&args.args)
+    }
+
+    fn aliases(&self) -> &[String] {
+        &self.aliases
+    }
+
+    fn coerce_types(
+        &self,
+        arg_types: &[DataType],
+    ) -> datafusion_common::Result<Vec<DataType>> {
+        if arg_types.len() != 1 {
+            return Err(invalid_arg_count_exec_err("expm1", (1, 1), arg_types.len()));
+        }
+        match &arg_types[0] {
+            DataType::Int64
+            | DataType::Utf8
+            | DataType::LargeUtf8
+            | DataType::Binary
+            | DataType::LargeBinary => Ok(vec![arg_types[0].clone()]),
+            DataType::Dictionary(key_type, value_type) => match value_type.as_ref() {
```

Comment:
@shehabgamin @alamb Is the correct way to handle dictionary-encoding coercion? I did not see this handled in the `expm1` function.

Reply:
I think it would be simpler if you could use some of the existing signatures and then return something for `signature()`:
https://github.com/apache/datafusion/blob/a39c07a2338a46fccc1111cafb5a2ffdcf6fd329/datafusion/functions/src/string/to_hex.rs#L116

That way the built in coercion logic would take over. If that doesn't work for spark functions, perhaps we can define some spark compatible signatures (as in implement `SparkSignature` that would return a https://docs.rs/datafusion/latest/datafusion/logical_expr/enum.TypeSignature.html)

Reply:
We can also wait for a few more PRs to see if there is a common pattern emerging before refactoring

Reply:
> @shehabgamin @alamb Is the correct way to handle dictionary-encoding coercion? I did not see this handled in the expm1 function.

FWIW I think it is the correct way to handle the dictionary coercion